File: java/ch17-04-arithmetic/src/test/java/com/thesoftwaregorilla/tdd/money/OperatorTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;

@DisplayName("Operation Class Tests")
public class OperatorTest {

    private Bank<Money> bank;
    @BeforeEach
    public void beforeEachSetUp() {
        bank = BankTest.getBankWithRates();
    }

    @AfterEach
    public void afterEachTearDown() {
        bank = new Bank<Money>();
    }

    @Test
    public void testSubtract() {
        Money usd = Money.from(new BigDecimal("5.00"), "USD", bank);
        Money zar = Money.from(new BigDecimal("17.00"), "ZAR", bank);
        Money result = usd.subtract(zar);
        assertInstanceOf(Money.class, result);
        assertEquals(Money.from(new BigDecimal("4.00"), "USD", bank), result);
    }

    @Test
    public void testAdd() {
        Money usd = Money.from(new BigDecimal("5.00"), "USD", bank);
        Money zar = Money.from(new BigDecimal("17.00"), "ZAR", bank);
        Money result = usd.add(zar);
        assertInstanceOf(Money.class, result);
        assertEquals(Money.from(new BigDecimal("6.00"), "USD", bank), result);
    }

    @Test
    public void testMultiply() {
        Money usd = Money.from(new BigDecimal("5.00"), "USD", bank);
        Money result = usd.multiply(new BigDecimal("2"));
        assertInstanceOf(Money.class, result);
        assertEquals(Money.from(new BigDecimal("10.00"), "USD", bank), result);
    }

    @Test
    public void testDivide() {
        Money usd = Money.from(new BigDecimal("5.00"), "USD", bank);
        Money result = usd.divide(new BigDecimal("2"));
        assertInstanceOf(Money.class, result);
        assertEquals(Money.from(new BigDecimal("2.50"), "USD", bank), result);
    }
}

File: cs/ch17-04-arithmetic/Money.Tests/OperatorTest.cs
namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion

[TestFixture]
[Description("Operator Tests")]
public class OperatorTest
{

    private Bank<Money> _bank;
    [SetUp]
    public void SetUp()
    {
        _bank = BankTest.GetBankWithRates();
    }

    [TearDown]
    public void TearDown()
    {
        _bank = new Bank<Money>();
    }

    [Test]
    public void TestSubtract() {
        Money usd = Money.From(5, "USD", _bank);
        Money zar = Money.From(17, "ZAR", _bank);
        Money result = usd.Subtract(zar);
        Assert.IsInstanceOf<Money>(result);
        Assert.That(result, Is.EqualTo(Money.From(4, "USD", _bank)));
        Assert.That(result, Is.EqualTo(usd - zar));
    }

    [Test]
    public void TestAdd() {
        Money usd = Money.From(5, "USD", _bank);
        Money zar = Money.From(17, "ZAR", _bank);
        Money result = usd.Add(zar);
        Assert.IsInstanceOf<Money>(result);
        Assert.That(result, Is.EqualTo(Money.From(6, "USD", _bank)));
        Assert.That(result, Is.EqualTo(usd + zar));

    }

    [Test]
    public void TestMultiply() {
        Money usd = Money.From(5, "USD", _bank);
        Money result = usd.Multiply(2);
        Assert.IsInstanceOf<Money>(result);
        Assert.That(result, Is.EqualTo(Money.From(10, "USD", _bank)));
        Assert.That(result, Is.EqualTo(usd * 2));
    }

    [Test]
    public void TestDivide() {
        Money usd = Money.From(5, "USD", _bank);
        Money result = usd.Divide(2);
        Assert.IsInstanceOf<Money>(result);
        Assert.That(result, Is.EqualTo(Money.From(2.5m, "USD", _bank)));
        Assert.That(result, Is.EqualTo(usd / 2));
    }

    [Test]
    public void TestEqualityOperators() {
        Money usd = Money.From(5, "USD", _bank);
        Money zar = Money.From(5, "ZAR", _bank);
        Assert.True(usd == Money.From(5, "USD", _bank));
        Assert.False(usd == Money.From(4, "USD", _bank));
        Assert.False(zar == usd);
        Assert.True(usd != zar);
        Assert.True(usd != Money.From(4, "USD", _bank));
        Assert.False(usd != Money.From(5, "USD", _bank));
    }
    
}
